spec/GraphQLOverHTTP.md

```diff
@@ -237,120 +311,304 @@ omitted.
+the client does not supply an `Accept` header, the server SHOULD treat the
+request as if it had `Accept: application/graphql+json`.
+
+Note: This default is in place to maintain compatibility with legacy clients.
+
+## Validation
+
+Validation of a well-formed GraphQL-over-HTTP request SHOULD apply all the
+[validation rules](https://spec.graphql.org/draft/#sec-Validation) specified by
+the GraphQL specification.
+
+The server MAY, at its discretion, apply additional validation rules.
 
-1. Disregard the `Accept` header and respond with the default content type of
-   `application/graphql+json`, specifying this in the `Content-Type` header; OR
-2. Respond with a `406 Not Acceptable` status code
+## Execution
 
-Note: For any non-`2XX` response, the client should not rely on the body to be
-in GraphQL format since the source of the response may not be the GraphQL server
-but instead some intermediary such as API gateways, firewalls, etc.
+Execution of a GraphQL-over-HTTP request follows
+[GraphQL's ExecuteRequest()](<https://spec.graphql.org/draft/#ExecuteRequest()>)
+algorithm.
 
 ## Status Codes
 
-If the response has Content-Type GraphQL and contains a non-null `data` entry,
-then it MUST have status code `2xx`, and it SHOULD have status code `200`
-(Okay).
+In case of errors that completely prevent the generation of a well-formed
+GraphQL response, the server SHOULD respond with the appropriate status code
+depending on the concrete error condition.
 
-If the response has Content-Type GraphQL and has a non-`2xx` status code, the
-`data` entry must be either:
+Note: Typically this will be the `400` (Bad Request) status code.
 
-- equal to `null`
-- not present
+Otherwise, the status codes depends on the media type with which the GraphQL
+response will be served:
+
+### application/json
+
+This section only applies when the response body is to use the
+`application/json` media type.
+
+The server SHOULD use the `200` status code, independent of any {GraphQL request
+error} or {GraphQL field error} raised.
+
+Note: A status code in the `4xx` or `5xx` ranges or status code `203` (and maybe
+others) could originate from intermediary servers; since the client cannot
+determine if an `application/json` response with arbitrary status code is a
+well-formed GraphQL response (because it cannot trust the source) the server
+must use `200` status code to guarantee to the client that the response has not
+been generated or modified by an intermediary.
+
+If the GraphQL response contains a non-null {data} entry then the server MUST
+use the `200` status code.
+
+Note: This indicates that no {GraphQL request error} was raised, though one or
+more {GraphQL field error} may have been raised this is still a successful
+execution - see "partial response" in the GraphQL specification.
+
+The server SHOULD NOT use a `4xx` or `5xx` status code.
+
+Note: For compatibility with legacy servers, this specification allows the use
```

Comment:
For what it's worth, Apollo Server does use 400s for GraphQL request errors (in a somewhat inconsistent way in AS3, and in a more consistent way in AS4). I don't think we are going to choose to change those to 200s for `application/json` as encouraged by this spec.  I do look forward to implementing `application/graphql+json` responses, though. My understanding is that this only violates SHOULDs, and I don't disagree with the recommendations of the spec.

Reply:
Not changing them is fine. Once 2025 rolls around this'll be a non-issue hopefully :crossed_fingers: